src/schema-analyzer.ts

```diff
@@ -339,89 +388,89 @@ export class SchemaAnalyzer {
-    if (typeName === 'Array') {
-      type.types = type.types ?? {};
-      type.lengths = type.lengths || [];
-      type.lengths.push(value.length);
-      value.forEach((v: SchemaBuildingMap) => this.addToType(path, v, type.types!));
-
-    // Recurse into nested documents by calling `addToField` for all sub-fields.
-    } else if (typeName === 'Document') {
-      type.fields = _.get(type, 'fields', {});
-      Object.entries(value).forEach(([k, v]) => this.addToField(`${path}.${k}`, v, type.fields!));
-
-    // If the `storeValues` option is enabled, store some example values.
-    } else if (this.options.storeValues) {
-      const defaultValue = bsonType === 'String'
-        ? Reservoir(100) : Reservoir(10000);
-      type.values = type.values || defaultValue;
-
-      addToValue(type, value);
-    }
-  }
-
-  /**
-   * Handles a field from a document. Passes the value to `addToType`.
-   *
-   * @param {String}  path     field path in dot notation
-   * @param {Any}     value    value of the field
-   * @param {Object}  schema   the updated schema object
-   */
-  addToField(path: string, value: any, schema: SchemaBuildingMap) {
-    const pathSplitOnDot = path.split('.');
-    const defaults = {
-      [path]: {
-        name: pathSplitOnDot[pathSplitOnDot.length - 1],
-        path: path,
-        count: 0,
-        types: {}
+  analyzeDoc(doc: Document) {
+    this.finalized = false;
+    /**
+     * Takes a field value, determines the correct type, handles recursion into
+     * nested arrays and documents, and passes the value down to `addToValue`.
+     * Note: This mutates the `schema` argument.
+     */
+    const addToType = (path: string, value: BSONValue, schema: SchemaAnalysisFieldTypes) => {
+      const bsonType = getBSONType(value);
+      // If semantic type detection is enabled, the type is the semantic type
+      // or the original bson type if no semantic type was detected. If disabled,
+      // it is always the bson type.
+      const typeName = (this.options.semanticTypes) ? this.getSemanticType(value, path) || bsonType : bsonType;
+      if (!schema[typeName]) {
+        schema[typeName] = {
+          name: typeName,
+          bsonType: bsonType,
+          path: path,
+          count: 0
+        };
+      }
+      const type = schema[typeName];
+      type.count++;
+
+      if (isArrayType(type)) {
+        // Recurse into arrays by calling `addToType` for each element.
+        type.types = type.types ?? {};
+        type.lengths = type.lengths ?? [];
+        type.lengths.push((value as BSONValue[]).length);
+        (value as BSONValue[]).forEach((v: BSONValue) => addToType(path, v, type.types));
+      } else if (isDocumentType(type)) {
+        // Recurse into nested documents by calling `addToField` for all sub-fields.
+        type.fields = type.fields ?? {};
+        Object.entries(value as Document).forEach(([k, v]) => addToField(`${path}.${k}`, v, type.fields));
```

Comment:
Do we want to handle the case of keys that contain dots here? I think the original version of this package just predates that feature.